# 1.PLANTEAMIENTO DEL PROBLEMA Y OBJETIVO DE LA INVESTIGACION

Los ataques suicidas terroristas representan una amenaza grave y multifacética debido a su alta letalidad, impacto psicológico y complejidad táctica. Entender sus características es crucial para desarrollar estrategias efectivas de prevención y respuesta. Esto incluye una combinación de inteligencia, seguridad física, programas de desradicalización y cooperación internacional para mitigar y, en última instancia, prevenir estos ataques devastadores. Son un fenómeno complejo con características distintivas que los diferencian de otros tipos de actos terroristas.
### Caracteristicas clave de los ataques terroristas suicidas :

### 1. *Motivación y Objetivos*
- *Sacrificio Personal*: Los perpetradores están dispuestos a morir en el ataque, lo que elimina la necesidad de escape y permite ejecutar misiones más arriesgadas y complejas.
- *Impacto Psicológico*: Estos ataques buscan causar terror y desmoralización en la población objetivo debido a su naturaleza letal y aparentemente irracional.
- *Simbología y Propaganda*: A menudo son utilizados por grupos terroristas para simbolizar el compromiso extremo con su causa y para reclutar y radicalizar a nuevos miembros.

### 2. *Técnicas y Tácticas*
- *Carga Explosiva Personal*: Los atacantes suelen llevar explosivos en su cuerpo o en objetos cercanos (como mochilas o cinturones explosivos).
- *Vehículos Bomba*: En algunos casos, los atacantes suicidas usan vehículos cargados con explosivos para maximizar el daño.
- *Ataques Coordinados*: A veces, múltiples atacantes suicidas actúan de manera coordinada para aumentar el impacto y la complejidad de la respuesta de seguridad.

### 3. *Seleccion de Objetivos*
- *Lugares Públicos*: Los atacantes suelen dirigirse a lugares con alta concentración de personas, como mercados, estaciones de transporte, conciertos, y eventos deportivos.
- *Objetivos Simbólicos*: También pueden elegir lugares que tienen un valor simbólico alto, como embajadas, edificios gubernamentales, instalaciones militares, y sitios religiosos.
- *Infraestructuras Críticas*: En ocasiones, los ataques se dirigen a infraestructuras críticas como aeropuertos, puertos y plantas de energía.

### 4. *Planificación y Ejecución*
- *Reclutamiento y Entrenamiento*: Los atacantes suicidas a menudo pasan por un proceso de radicalización y entrenamiento intensivo, tanto psicológico como técnico.
- *Logística y Apoyo*: Requieren una red de apoyo para proporcionar explosivos, transporte, inteligencia sobre el objetivo y, en algunos casos, refugio antes del ataque.
- *Innovación Táctica*: Los métodos y las técnicas evolucionan para adaptarse a las medidas de seguridad, utilizando tecnologías y tácticas que maximicen el éxito del ataque.

### 5. *Perfil de los Atacantes*
- *Diversidad Demográfica*: Los atacantes pueden provenir de diversas edades, géneros y orígenes socioeconómicos, aunque comúnmente son jóvenes.
- *Motivaciones Ideológicas*: Las motivaciones suelen estar profundamente arraigadas en creencias ideológicas, religiosas o políticas extremas.
- *Estado Mental*: Los perpetradores pueden estar sujetos a adoctrinamiento intenso, manipulación emocional y presión social, que los lleva a ver el ataque como un acto de martirio o sacrificio heroico.

### 6. *Impacto y Consecuencias*
- *Alta Letalidad*: Estos ataques suelen resultar en un alto número de víctimas mortales y heridos debido a la proximidad del atacante a los objetivos.
- *Desestabilización Social y Política*: Pueden desencadenar respuestas desproporcionadas de las fuerzas de seguridad y generar tensión política y social.
- *Cobertura Mediática*: La naturaleza extrema y dramática de estos ataques atrae una amplia cobertura mediática, amplificando su impacto psicológico y propagandístico.

### 7. *Prevención y Contramedidas*
- *Inteligencia y Monitoreo*: La identificación temprana de redes de radicalización y la vigilancia de potenciales atacantes son cruciales.
- *Seguridad Física*: Mejora de la seguridad en lugares públicos y objetivos potenciales mediante controles de acceso y dispositivos de detección de explosivos.
- *Programas de Desradicalización*: Iniciativas para prevenir la radicalización y reintegrar a individuos que han sido adoctrinados.



Prevenir ataques terroristas suicidas es crucial no solo por la protección directa de vidas humanas, sino también por sus beneficios en términos de seguridad nacional, estabilidad social, bienestar psicológico, impacto económico, y cooperación internacional. Las medidas preventivas efectivas son fundamentales para mantener la paz y la seguridad en las sociedades modernas.
###Importancia de su estudio :

### 1. *Protección de Vidas Humanas*
- *Minimización de Pérdidas*: Los ataques suicidas suelen ser altamente letales, causando un gran número de muertes y heridas graves en un solo evento. Prevenir estos ataques directamente salva vidas y reduce el número de víctimas.
- *Prevención de Daños Colaterales*: Además de las víctimas directas, los ataques pueden causar daños colaterales significativos, afectando a personas cercanas al lugar del ataque.

### 2. *Estabilidad y Seguridad Nacional*
- *Prevención de la Desestabilización*: Los ataques suicidas pueden desestabilizar regiones enteras, creando caos y desconfianza en las instituciones de seguridad.
- *Fortalecimiento de la Confianza Pública*: La capacidad de prevenir ataques refuerza la confianza del público en el gobierno y las fuerzas de seguridad, contribuyendo a la estabilidad social y política.

### 3. *Impacto Psicológico y Social*
- *Reducción del Miedo y la Ansiedad*: Los ataques terroristas, especialmente los suicidas, tienen un impacto psicológico profundo, causando miedo y ansiedad generalizados en la población. Prevenir estos ataques ayuda a mantener la moral pública y la tranquilidad.
- *Cohesión Social*: Evitar ataques terroristas ayuda a prevenir divisiones sociales y étnicas que los terroristas intentan explotar para crear conflictos internos.

### 4. *Impacto Económico*
- *Protección de Infraestructura*: Los ataques suicidas a menudo apuntan a infraestructuras críticas, como transportes públicos, instalaciones gubernamentales y centros comerciales. La prevención protege estos activos esenciales.
- *Reducción de Costos*: El impacto económico de un ataque terrorista puede ser devastador, incluyendo costos de emergencia, reparación de daños, y pérdidas económicas por la interrupción de negocios y el turismo. La prevención ahorra estos costos.

### 5. *Prevención de Radicalización*
- *Interrupción de Cadenas de Reclutamiento*: La prevención de ataques incluye esfuerzos para desmantelar redes terroristas y programas de desradicalización, reduciendo la probabilidad de que individuos sean reclutados para cometer ataques suicidas.
- *Educación y Concientización*: Las campañas preventivas también pueden incluir programas educativos que disuadan a potenciales terroristas suicidas, abordando las raíces de la radicalización.

### 6. *Cooperación Internacional*
- *Fortalecimiento de la Colaboración*: La prevención efectiva de ataques suicidas requiere cooperación internacional, compartiendo inteligencia y estrategias entre países, fortaleciendo alianzas globales contra el terrorismo.
- *Cumplimiento de Obligaciones Globales*: Participar en la prevención de ataques terroristas ayuda a los países a cumplir con sus responsabilidades en materia de seguridad global y derechos humanos.



In [65]:
#Importar librerias necesarias
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
!pip install prophet
from prophet import Prophet
from sklearn.preprocessing import MinMaxScaler
from sklearn.model_selection import train_test_split
from sklearn.feature_selection import f_regression, SelectKBest
from sklearn.metrics import accuracy_score
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score, roc_auc_score, confusion_matrix
from sklearn.ensemble import RandomForestClassifier, GradientBoostingClassifier
from xgboost import XGBClassifier
from sklearn.metrics import classification_report, confusion_matrix
from sklearn.tree import DecisionTreeClassifier, DecisionTreeRegressor
from sklearn.ensemble import RandomForestClassifier, RandomForestRegressor
from sklearn.ensemble import GradientBoostingClassifier, GradientBoostingRegressor
import xgboost as xgb
from xgboost import XGBClassifier, XGBRegressor



# 2.EDA

# Limpieza variables sobrantes

In [66]:
#  leer el archivo CSV con diferentes delimitadores
try:
    data = pd.read_csv("globalterrorismdb.csv", encoding='latin1')
except pd.errors.ParserError:
    data = pd.read_csv("../data/raw/globalterrorismdb.csv", sep=",", encoding='latin1')
except pd.errors.ParserError:
    data = pd.read_csv("../data/raw/globalterrorismdb.csv.csv", sep="\t", encoding='latin1')

pd.set_option('display.max_columns', None)

# Mostrar las primeras filas del dataframe para verificar que se ha cargado correctamente
data.head()

eventid  iyear  imonth  iday approxdate  extended resolution  country  \
0  197000000001   1970       7     2        NaN         0        NaN       58   
1  197000000002   1970       0     0        NaN         0        NaN      130   
2  197001000001   1970       1     0        NaN         0        NaN      160   
3  197001000002   1970       1     0        NaN         0        NaN       78   
4  197001000003   1970       1     0        NaN         0        NaN      101   

          country_txt  region                   region_txt provstate  \
0  Dominican Republic       2  Central America & Caribbean       NaN   
1              Mexico       1                North America   Federal   
2         Philippines       5               Southeast Asia    Tarlac   
3              Greece       8               Western Europe    Attica   
4               Japan       4                    East Asia   Fukouka   

            city   latitude   longitude  specificity  vicinity location  \
0  Santo Domingo  18.456792  -69.951164          1.0         0      NaN   
1    Mexico city  19.371887  -99.086624          1.0         0      NaN   
2        Unknown  15.478598  120.599741          4.0         0      NaN   
3         Athens  37.997490   23.762728          1.0         0      NaN   
4        Fukouka  33.580412  130.396361          1.0         0      NaN   

  summary  crit1  crit2  crit3  doubtterr  alternative alternative_txt  \
0     NaN      1      1      1        0.0          NaN             NaN   
1     NaN      1      1      1        0.0          NaN             NaN   
2     NaN      1      1      1        0.0          NaN             NaN   
3     NaN      1      1      1        0.0          NaN             NaN   
4     NaN      1      1      1       -9.0          NaN             NaN   

   multiple  success  suicide  attacktype1                 attacktype1_txt  \
0       0.0        1        0            1                   Assassination   
1       0.0        1        0            6     Hostage Taking (Kidnapping)   
2       0.0        1        0            1                   Assassination   
3       0.0        1        0            3               Bombing/Explosion   
4       0.0        1        0            7  Facility/Infrastructure Attack   

   attacktype2 attacktype2_txt  attacktype3 attacktype3_txt  targtype1  \
0          NaN             NaN          NaN             NaN         14   
1          NaN             NaN          NaN             NaN          7   
2          NaN             NaN          NaN             NaN         10   
3          NaN             NaN          NaN             NaN          7   
4          NaN             NaN          NaN             NaN          7   

                 targtype1_txt  targsubtype1  \
0  Private Citizens & Property          68.0   
1      Government (Diplomatic)          45.0   
2          Journalists & Media          54.0   
3      Government (Diplomatic)          46.0   
4      Government (Diplomatic)          46.0   

                                    targsubtype1_txt  \
0                                     Named Civilian   
1  Diplomatic Personnel (outside of embassy, cons...   
2                    Radio Journalist/Staff/Facility   
3                                  Embassy/Consulate   
4                                  Embassy/Consulate   

                         corp1                  target1  natlty1  \
0                          NaN             Julio Guzman     58.0   
1  Belgian Ambassador Daughter  Nadine Chaval, daughter     21.0   
2             Voice of America                 Employee    217.0   
3                          NaN             U.S. Embassy    217.0   
4                          NaN           U.S. Consulate    217.0   

          natlty1_txt  targtype2 targtype2_txt  targsubtype2 targsubtype2_txt  \
0  Dominican Republic        NaN           NaN           NaN              NaN   
1             Belgium        NaN           NaN           NaN              NaN   


In [67]:
print(f"Estructura del dataset: {data.shape}")
print("--------------------------------------")
data.info()

Estructura del dataset: (181691, 135)
--------------------------------------
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 181691 entries, 0 to 181690
Columns: 135 entries, eventid to related
dtypes: float64(55), int64(22), object(58)
memory usage: 187.1+ MB


Este dataset contiene informacion detallada sobre más de 180,000 casos de indicentes terroristas, abarcando una amplia gama de variables clave como son la fecha y ubicacion del incidente, armas usadas, naturaleza del objetivo, número de víctimas...

In [68]:
#!Revisar eventos repetidos, esto da igual pk va to a la basura
valores_duplicados = data[data.duplicated(subset='eventid', keep = False)]
print(f"Valores duplicados: {valores_duplicados}")

Valores duplicados: Empty DataFrame
Columns: [eventid, iyear, imonth, iday, approxdate, extended, resolution, country, country_txt, region, region_txt, provstate, city, latitude, longitude, specificity, vicinity, location, summary, crit1, crit2, crit3, doubtterr, alternative, alternative_txt, multiple, success, suicide, attacktype1, attacktype1_txt, attacktype2, attacktype2_txt, attacktype3, attacktype3_txt, targtype1, targtype1_txt, targsubtype1, targsubtype1_txt, corp1, target1, natlty1, natlty1_txt, targtype2, targtype2_txt, targsubtype2, targsubtype2_txt, corp2, target2, natlty2, natlty2_txt, targtype3, targtype3_txt, targsubtype3, targsubtype3_txt, corp3, target3, natlty3, natlty3_txt, gname, gsubname, gname2, gsubname2, gname3, gsubname3, motive, guncertain1, guncertain2, guncertain3, individual, nperps, nperpcap, claimed, claimmode, claimmode_txt, claim2, claimmode2, claimmode2_txt, claim3, claimmode3, claimmode3_txt, compclaim, weaptype1, weaptype1_txt, weapsubtype1, weapsubtyp

Tras comprobar si encontramos valores repetidos toca comprobar los valores nulos de cada columna del dataset

In [ ]:
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

# Valores nulos por columna
valores_nulos = data.isnull().sum()
print(f"Número de valores nulos por columna:\n{valores_nulos}")

A continuación, procedemos a eliminar la columna eventID debido a que solo sirve como identificador y hemos verificado que no contiene valores duplicados. Además de esta columna hemos identificado otras columnas que vamos a eliminar debido a la gran cantidad de valores perdidos que ocupan y de su baja relevancia para nuestro análisis

In [ ]:
#Listado de columnas a eliminar
eliminar_columnas = [
    'specificity', 'multiple',"individual", 'target1', 'propvalue', 'corp1', 'longitude', 'latitude',
     'specifity', 'vicinity', 'doubtterr',
    'weapsubtype1_txt', 'nperps', 'nperpcap', 'claimed', 'nkillter', 'nwoundte',
    'eventid', 'location', 'summary', 'alternative', 'alternative_txt', 'attacktype2', 'attacktype2_txt',
    'attacktype3', 'attacktype3_txt', 'targtype2', 'targtype2_txt', 'targsubtype2', 'targsubtype2_txt',
    'corp2', 'target2', 'natlty2', 'natlty2_txt', 'targtype3', 'targtype3_txt', 'targsubtype3',
    'targsubtype3_txt', 'corp3', 'target3', 'natlty3', 'natlty3_txt', 'gsubname', 'gname2', 'gsubname2',
    'gname3', 'gsubname3', 'guncertain2', 'guncertain3', 'claim2', 'claimmode2', 'claimmode2_txt',
    'claim3', 'claimmode3', 'claimmode3_txt', 'compclaim', 'weaptype2', 'weaptype2_txt', 'weapsubtype2',
    'weapsubtype2_txt', 'weaptype3', 'weaptype3_txt', 'weapsubtype3', 'weapsubtype3_txt', 'weaptype4',
    'weaptype4_txt', 'weapsubtype4', 'weapsubtype4_txt','addnotes', 'scite1', 'scite2', 'scite3', 'propextent',
    'propextent_txt', 'ransomamt', 'ransomamtus', 'ransompaid',
    'ransompaidus', 'extended', 'country', 'region', 'crit1',
    'crit2', 'crit3', 'attacktype1', 'targtype1', 'targsubtype1',
    'natlty1', 'motive', 'guncertain1', 'weaptype1',
    'weapsubtype1', 'weapdetail', 'nkillus', 'nwoundus',
    'nhostkidus', 'divert', 'ransomnote', 'hostkidoutcome',
    'hostkidoutcome_txt', 'dbsource', 'related', 'propcomment','approxdate',
    'resolution', 'targsubtype1_txt','claimmode','claimmode_txt',
    'nhostkid','nhours','ndays','kidhijcountry','ransom','nreleased', 'INT_ANY', 'INT_MISC'
    ]

#Eliminamos las columnas sobrantes
data = data.drop(columns = eliminar_columnas, errors = 'ignore')

#Volvemos a imprimir el dataset para verificar las nuevas columnas
data.head()


Tras eliminar las columnas innecesarias optimizamos las que tenemos, vamos a unir las columnas iyear,imonth e iday en 1 sola y además procedemos a renombrar las restantes para un mejor entendimiento

In [ ]:
fecha = data[['iyear', 'imonth', 'iday']]

def valores_unicos(data, columnas):
    """
    Esta funcion nos permite obtener los valores unicos de las columnas indicadas.

    Parámetros:
    data: El dataset que contiene los datos.
    columnas(list): Lista de nombres de las columnas que vamos a obtener valores unicos

    Devuelve:
    Un dataset que contiene las columnas especificadas y solo los valores unicos de cada una
    """

    data_unicos = data[columnas].drop_duplicates()
    return data_unicos

data = pd.DataFrame(data)
columnas_unicas = valores_unicos(data, ['iyear', 'imonth', 'iday'])

print(columnas_unicas)

Podemos apreciar que hay valores "0" en la columna de imonth y iday, segun la documentacion encontrada online, entendemos que para los casos en los cuales no se tiene el dato certero o no ha recopilado el valor usado es "0". Procederemos con la sustitución de los valores "0" con la mediana, ya que proporciona una medida de tendencia central más robusta.

Dado esto, vamos a sustituir los valores "0" por la mediana

In [ ]:
mediana_iday = data['iday'].median()
mediana_imonth = data['imonth'].median()

data['iday'] = data['iday'].replace(0, mediana_iday)
data['imonth'] = data['imonth'].replace(0, mediana_imonth)

Tras sustituir estos valores es hora de reemplazar estas 3 columnas y unificarlas todas en 1 sola

In [ ]:
data['date'] = pd.to_datetime(data[['iday', 'imonth', 'iyear']].rename(columns={'iyear': 'year', 'imonth': 'month', 'iday': 'day'}), format='%d-%m-%Y')

#Tras la sustitucion eliminamos las 3 columnas usadas y volvemos a imprimir el dataset para comprobar
data.drop(['iyear', 'imonth', 'iday'], axis=1, inplace=True)
data.head()

**ANALISIS DE ATAQUES SUICIDAS EN EL** **TIEMPO** bold text

In [ ]:
# VAmos a ver los ataques suicidas en particular, primero por ano
data['year'] = data['date'].dt.year
# filtrmaos solo suicidas
suicide_attacks = data[data['suicide'] == 1]
# Agregar los datos por ano
yearly_suicide_attacks = suicide_attacks.groupby('year').size().reset_index(name='count')
# show
print(yearly_suicide_attacks)

Se puede comprobar la importancia de estudiar en profundidad los ataques suicidas pues hay un gran incremento del numero de casos en los ultimos años.

In [ ]:
#VAMOS A GRAFICAR   ATAQUES SUICIDE
import matplotlib.pyplot as plt
plt.figure(figsize=(10, 6))
plt.plot(yearly_suicide_attacks['year'], yearly_suicide_attacks['count'], marker='o')
plt.title('Suicide Attacks per Year')
plt.xlabel('Year')
plt.ylabel('Number of Suicide Attacks')
plt.grid(True)
plt.show()

Podemos ver que el primer ataque suicida registrado en el dataset data de 1981 (los registros inician en 1970), por lo que el uso de este tipo de tactica se comienza a registrar recien en el mencionado año.
Desde el año 2000, se observa un incremento notable, con un incremento de 37 en 2000 a 51 en 2001. Entendemos que este incremento puede estar vinculado al 9/11 y sus repercusiones. Facilmente podemos distinguir el crecimiento constante de los ataques, con un pico de 985 en 2016. El numero de ataques disminuye ligeramente en 2017, siendo el primer año en que la tendencia es negativa (analizando desde 2010).  
El metodo suicida se ha vuelto atractivo para los grupos terroristas debido a su efectividad para generar panico, obtener visibilidad de su causa y cumplir con sus objetivos (ideologicos, politicos).


TIMESERIES CON PROPHET
 **Usaremso el modelo prophet de facebook**
https://facebook.github.io/prophet/docs/quick_start.html#python-api

In [ ]:
!pip install prophet
import pandas as pd
import matplotlib.pyplot as plt
from prophet import Prophet

In [ ]:
#tomamos la lista ya creada de ataques suicidas por año y cambiamos nombre de col para Propeht
df = yearly_suicide_attacks.rename(columns={'year': 'ds', 'count': 'y'})

# Convertir el año a datetime
df['ds'] = pd.to_datetime(df['ds'], format='%Y')

# Verificar el DataFrame
print(df) #convertir a datetime

In [ ]:
#creamos y entrenamos
modelo = Prophet(yearly_seasonality=True)
modelo.fit(df)

#se crea dataframe que guarda las fecahs futuras y hacemeos la prediccion --- 10 años
future = modelo.make_future_dataframe(periods=10, freq='Y')
forecast = modelo.predict(future)

#graficamos
fig1 = modelo.plot(forecast)
plt.title('Predicción de Ataques Suicidas para los Próximos 10 años')
plt.xlabel('Año')
plt.ylabel('Número de Ataques Suicidas')
plt.show()

In [ ]:
print(forecast[['ds', 'yhat', 'yhat_lower', 'yhat_upper']].tail(10)) #IMPRIMIMOS EL FORECAST

El modelo y su prediccion sugieren una tendencia continua de incremento en la actividad de ataques suicidas en el futuro cercano, acompanado de los intervalos de confianza que reflejan la incertidumbre inherente a las predicciones. Los numeros pueden cambiar bruscamente si los grupos implementan metodos aun mas mortiferos o por el contrario, si las medidas de seguridad y prevención se vuelven mas duras.

Elegir los datos de los últimos tres años en el dataset de terrorismo global tiene tanto ventajas como inconvenientes:

Ventajas:

Relevancia Actual:

**Tendencias Recientes: Los datos más recientes reflejan las tendencias y patrones actuales del terrorismo, lo cual puede ser crucial para la toma de decisiones y políticas actuales.
Cambios en Tácticas: Las tácticas, objetivos y métodos de los grupos terroristas pueden cambiar con el tiempo. Usar datos recientes ayuda a capturar estas evoluciones.
Datos Más Consistentes:

*Mejora en la Recolección de Datos: La calidad de los datos y las metodologías de recolección pueden haber mejorado con el tiempo, haciendo que los datos recientes sean más precisos y completos.
Contexto Político Actual:

*Políticas y Eventos Recientes: Los datos recientes tienen más relevancia en el contexto de las políticas y eventos geopolíticos actuales. Esto puede ser vital para análisis específicos y respuestas rápidas.

Inconvenientes:

Volumen de Datos Reducido:

*Menor Cantidad de Datos: Utilizar solo los últimos tres años reduce significativamente la cantidad de datos disponibles, lo cual puede afectar la robustez de los modelos estadísticos y de machine learning.
Menor Diversidad de Casos: Con menos datos, hay menos variabilidad en los casos de estudio, lo que podría limitar la capacidad de generalización del análisis.
Perdida de Tendencias a Largo Plazo:

*Análisis de Tendencias a Largo Plazo: No se pueden identificar ni analizar tendencias a largo plazo o ciclos en la actividad terrorista, lo cual puede ser importante para comprender el fenómeno en su totalidad.

*Estabilidad de Patrones: Algunas conclusiones importantes sobre la estabilidad o los cambios graduales en los patrones de terrorismo pueden perderse.

*Sesgo Temporal:

Eventos Recientes Pueden Ser Atípicos: Los últimos años pueden incluir eventos excepcionales (por ejemplo, cambios abruptos debido a una pandemia global) que podrían sesgar los análisis y no reflejar patrones típicos a largo plazo.

*Falta de Contexto Histórico: Sin datos históricos, es difícil contextualizar eventos recientes y entender si son parte de una tendencia mayor o simplemente anomalías.

Ya que el objetivo es predecir futuros ataques basándose en tácticas y patrones recientes, los datos recientes son más apropiados.
Otro método posterior a utilizar seria donde se empleen datos recientes para capturar tendencias actuales, complementados con datos históricos para asegurar robustez y contexto histórico.

In [ ]:
# Filtrar los datos para los años 2015, 2016 y 2017
data_2015 = data[(data['date'].dt.year >= 2015) & (data['date'].dt.year <= 2017)]

# Verificar el resultado
print(data_2015.head())
print(data_2015['date'].dt.year.unique())

In [ ]:
data_2015.shape

Ahora pasamos a tener casi 40.000 filas y 19 columnas.

In [ ]:
# Obtener los valores únicos de la columna 'nkill' y ordenarlos en orden descendente
unique_nkill_sorted = data_2015['nkill'].dropna().unique()
unique_nkill_sorted.sort()
top_25_nkill = unique_nkill_sorted[-25:]

# Crear un diccionario para almacenar los valores más altos de 'nkill' y sus fechas asociadas
top_25_nkill_dates = {}

# Rellenar el diccionario con los valores únicos y sus fechas asociadas
for value in top_25_nkill:
    dates = data_2015.loc[data_2015['nkill'] == value, 'date']
    top_25_nkill_dates[value] = dates.unique()

# Mostrar los resultados
for nkill_value, dates in top_25_nkill_dates.items():
    print(f"nkill: {nkill_value}")
    for date in dates:
        print(f"  Date: {date}")

In [ ]:
# Obtener la primera y la última fecha
min_date = data_2015['date'].min()
max_date = data_2015['date'].max()

# Mostrar los resultados
print(f"La primera fecha es: {min_date}")
print(f"La última fecha es: {max_date}")

#Analisis variables categóricas

El primer análisis se centrará en las variables categóricas. Primero identificaremos cuáles son estas variables y luego contaremos el número de instancias únicas en cada una para analizarlas visualmente a través de bar plots

In [ ]:
'variables_cat = data_2015.select_dtypes(include=['object']).columns

print("Variables categóricas y sus valores únicos:")
for var in variables_cat:
    unique_values = data_2015[var].nunique()
    print(f"{var}: {unique_values} valores únicos")

In [ ]:

unique_attacktype1_txt = data['attacktype1_txt'].unique()
unique_weaptype1_txt = data['weaptype1_txt'].unique()
unique_targtype1_txt = data['targtype1_txt'].unique()

# Mostrar los valores únicos
print("Valores únicos de 'attacktype1_txt':")
print(unique_attacktype1_txt)

print("\nValores únicos de 'weaptype1_txt':")
print(unique_weaptype1_txt)

print("\nValores únicos de 'targtype1_txt':")
print(unique_targtype1_txt)


Como podemos ver, algunas variables tienen demasiados valores únicos para mostrarlos adecuadamente con un bar plot. Mostraremos las 15 principales categorías en cada caso para que sea visualmente legible y analizable

In [ ]:

print(data_2015['date'].dt.year.unique())

In [ ]:
# modificamos los nombres de las variables para los graficos
import matplotlib.pyplot as plt
title_mapping = {
    'city': 'Ciudades donde ocurrieron los incidentes',
    'country_txt': 'Países donde ocurrieron los incidentes',
    'region_txt': 'Regiones donde ocurrieron los incidentes',
    'targtype1_txt': 'Tipos de objetivo',
    'corp1': 'Nombre de entidades objetivo',
    'natlty1_txt': 'Nacionalidad de los objetivos principales',
    'attacktype1_txt': 'Tipos de ataque',
    'weaptype1_txt': 'Tipos de armas utilizadas en el ataque',
    'gname': 'Nombres de grupos terroristas',
}

data_2015_suicide = data[(data['date'].dt.year >= 2015) & (data['date'].dt.year <= 2017) & (data['suicide'] == 1)]

# Seleccionamos las columnas categóricas
categorical_variables = data_2015_suicide.select_dtypes(include=['object']).columns

# Graficos de barra para poder ver nuestras variables
for var in categorical_variables:
    if var in data_2015_suicide.columns:
        plt.figure(figsize=(10, 6))
        top_10_counts = data_2015_suicide[var].value_counts().nlargest(10)

        # Truncamos los nombres a 30 caracteres para que se pueda leer
        top_10_counts.index = top_10_counts.index.str[:30]
        top_10_counts.plot(kind='bar', color='grey')

        title = title_mapping.get(var, var)
        plt.title(f'Top 10: {title}')

        plt.xlabel(var)
        plt.ylabel('Frecuencia')
        plt.xticks(rotation=45, ha='right')
        plt.tight_layout()
        plt.show()

###ANALISIS DE GRAFICOS

Podemos ver que Iraq es donde ocurrieron la mayoría de los ataques suicidas, con casi 1200, seguido por Afganistán y Nigeria  (ambos alrededor de 400 y 300 respectivamente). Esto concuerda con lo mostrado en el gráfico de regiones,  estados, ciudades). Las provincias de Al Anbar y Nineveh (Iraq) son las mas afectadas por este tipo de ataques, seguidas por Borno (Nigeria).

Tipos de Ataques y armas:
Mas de 2500 ataques fueron perpetrados con bombas o explosiones.  Hay significativamente menos casos de  armas de fuego, toma de rehenes y ataques a instalaciones. También hubo pocos asaltos sin armas y secuestros.

Tipos de Objetivos:
Al observar los tipos de objetivos, vemos que más de 800 afectaron a objetivos militares, seguido por mas de 400 a ciudades y propiedades, y alrededor de 400 a policías. Hay pocos casos de estaciones diplomáticas, transportes, etc.

Grupos Terroristas:
ISIL/ISIS (Iraq) es el grupo que mas ataques suicida ha cometido.  Hay alrededor de 500 casos para los cuales se desconoce el grupo terrorista al que pertenecía/n los atacantes. Boko Haram (Nigeria) sigue en la lista con alrededor de 400 ataques suicidas — este grupo también es conocido como ISIS WA (West Africa). Los talibanes (Afganistán) aparecen en cuarto lugar con alrededor de 200 incidentes suicidas.


A continuacion, graficamos histogramas para las variables categoricas que tienen pocos unique values, a saber en este caso, SI, NO y Desconocido

In [ ]:
# List of binary/categorical columns
binary_cols = ['success', 'property','ishostkid','INT_LOG', 'INT_IDEO']


# Mapeo de títulos
title_mapping2 = {
    'success': 'Ataque fue exitoso',
    'property': 'Propiedad dañada',
    'ishostkid': 'Rehenes',
    'INT_LOG': 'Ataque logístico internacional',
    'INT_IDEO': 'Ataque ideológico internacional',
}

# Para que aparezcan codifico los valores
value_mapping = {
    -9: 'Desconocido',
    0: 'No',
    1: 'Sí'
}

# Plot histograms for binary/categorical columns
for col in binary_cols:
    plt.figure(figsize=(10, 6))
    # Mapear los valores numéricos a etiquetas antes de trazar el histograma
    mapped_data = data_2015_suicide[col].map(value_mapping)
    mapped_data.value_counts().plot(kind='bar', color='grey')

    # Usar el título mapeado si está disponible, de lo contrario usar el nombre original de la variable
    title1 = title_mapping2.get(col, col)
    plt.title(f'Grafico de {title1}')
    plt.xlabel(col)
    plt.ylabel('Frecuencia')
    plt.xticks(rotation=45, ha='right')
    plt.tight_layout()
    plt.show()

**Vamos a analizar las variables binarias**.
Comenzamos por el éxito o no del ataque. Vemos que 2000 de los ataques fueron exitosos, mientras mas de 700 no lo fueron. Esto se remite a que por ejemplo, un ataque con bombas o explosivos pudo haber sucedido sin haber causado muertes, heridos o danos a la propiedad.
Analizando el daño a la propiedad, observamos que en mas de 1200 de los incidentes no hubo daños material a la propiedad, mientras que en menos de 800 casos si hubo (este dato es consistente con el de suceso del ataque). Hay poco mas de 600 casos donde el daño a la propiedad es desconocido.
En casi 100% de los casos no hubo rehenes retenidos.
Yendo al dato logístico del ataque, vemos que 2000 de los mismo han sucedido domésticamente, es decir,  los perpetradores no han cruzado fronteras para llevar a cabo el ataque.


In [ ]:
#Convertir -9 desconocido, 0 a NO y 1 a SI
columns_to_convert = ['success', 'suicide', 'property', 'ishostkid', 'INT_LOG', 'INT_IDEO']

def convert_values(value):
    if value == -9:
        return 'Desconocido'
    elif value == 1:
        return 'Sí'
    elif value == 0:
        return 'No'
    else:
        return value


# Convertimos las columnas especificadas
for column in columns_to_convert:
    data_2015[column] = data_2015[column].apply(convert_values)

print(data_2015.head())

#Analisis variables numericas

# Ingenieria caracterisiticas

In [ ]:
data_2015.describe()

In [ ]:
data_2015.columns


In [ ]:
data.info()

In [ ]:
variables = ['country_txt', 'region_txt', 'provstate', 'city', 'success', 'suicide',
       'attacktype1_txt', 'targtype1_txt', 'natlty1_txt', 'gname',
       'weaptype1_txt', 'nkill', 'nwound', 'property', 'ishostkid', 'INT_LOG',
       'INT_IDEO', 'date', 'year']

# Convertir variables categóricas a numéricas
categorical_vars = ['country_txt', 'region_txt', 'provstate', 'city', 'success', 'suicide',
       'attacktype1_txt', 'targtype1_txt', 'natlty1_txt', 'gname',
       'weaptype1_txt', 'property', 'ishostkid', 'INT_LOG',
       'INT_IDEO']

for var in categorical_vars:
    data_2015[var + '_num'] = pd.factorize(data_2015[var])[0]

# Filtrar solo las columnas numéricas
numeric_variables = [var for var in data_2015.columns if var in variables or var.endswith('_num')]
numeric_variables = [var for var in numeric_variables if pd.api.types.is_numeric_dtype(data_2015[var])]

# Definir el tamaño de la figura
plt.figure(figsize=(20, 40))

# Calcular el número de filas y columnas
num_vars = len(numeric_variables)
cols = 5
rows = (num_vars // cols) + 1

# Crear subplots
fig, axes = plt.subplots(rows, cols, figsize=(20, 40))
axes = axes.flatten()

for i, var in enumerate(numeric_variables):
    sns.boxplot(data_2015[var], ax=axes[i])
    axes[i].set_title(var)

# Eliminar subplots vacíos
for j in range(i + 1, len(axes)):
    fig.delaxes(axes[j])

# Ajustar layout
plt.tight_layout()
plt.show()

Tras graficas todas las columnas calculamos los límites de los outliers para determinar que valores están fuera de rango

In [ ]:
def calculate_outliers_limits(data_2015, column):
    stats = data_2015[column].describe()
    iqr = stats["75%"] - stats["25%"]
    upper_limit = stats["75%"] + 1.5 * iqr
    lower_limit = stats["25%"] - 1.5 * iqr
    return upper_limit, lower_limit, iqr

# Iterar sobre las columnas numéricas y calcular los límites
outliers_limits = {}

for col in numeric_variables:
    upper, lower, iqr = calculate_outliers_limits(data_2015, col)
    outliers_limits[col] = {
        "upper_limit": round(upper, 2),
        "lower_limit": round(lower, 2),
        "iqr": round(iqr, 2)
    }

# Mostrar los resultados
for col, limits in outliers_limits.items():
    print(f"Límites para la variable {col}:")
    print(f"  - Límite superior: {limits['upper_limit']}")
    print(f"  - Límite inferior: {limits['lower_limit']}")
    print(f"  - Rango intercuartílico: {limits['iqr']}")
    print()

In [ ]:
# Función para reemplazar outliers con la mediana
def replace_outliers_with_median(data_2015, column, lower_bound, upper_bound):
    median = data_2015[column].median()
    data_2015[column] = data_2015[column].apply(lambda x: median if x < lower_bound or x > upper_bound else x)

# Tratamiento de datos atípicos
replace_outliers_with_median(data_2015, 'nkill', -3.0, 5.0)
replace_outliers_with_median(data_2015, 'nwound', -3.0, 5.0)
replace_outliers_with_median(data_2015, 'country_txt_num', -55.0, 161.0)
replace_outliers_with_median(data_2015, 'region_txt_num', -2.5, 17.5)
replace_outliers_with_median(data_2015, 'provstate_num', -938.5, 2097.5)
replace_outliers_with_median(data_2015, 'city_num', -23125.0, 40171.0)
replace_outliers_with_median(data_2015, 'attacktype1_txt_num', -1.0, 7.0)
replace_outliers_with_median(data_2015, 'targtype1_txt_num', -9.5, 18.5)
replace_outliers_with_median(data_2015, 'natlty1_txt_num', -37.5, 142.5)
replace_outliers_with_median(data_2015, 'gname_num', -2185.0, 3647.0)
replace_outliers_with_median(data_2015, 'weaptype1_txt_num', -2.0, 6.0)

# El tratamiento de datos atípicos en variables binarias ya no sería necesario pues solo hay valores si, no y desconocido en su defecto

# Verificación final
print("Tratamiento de datos atípicos completado.")
print(data_2015.describe())


In [ ]:
data_2015.shape

In [ ]:
data_2015.columns

In [ ]:
# Lista de columnas a eliminar
columnas_a_eliminar = ['country_txt', 'region_txt', 'provstate', 'city', 'success', 'suicide',
                       'attacktype1_txt', 'targtype1_txt', 'natlty1_txt', 'gname',
                       'weaptype1_txt', 'property', 'ishostkid', 'INT_LOG',
                       'INT_IDEO', 'date', 'year']

# Eliminar las columnas del DataFrame
data_2015 = data_2015.drop(columns=columnas_a_eliminar)



In [ ]:
print(data_2015.columns)

In [ ]:
# Calcular la mediana de las columnas 'nkill' y 'nwound'
median_nkill = data_2015['nkill'].median()
median_nwound = data_2015['nwound'].median()

# Imputar los valores perdidos con la mediana
data_2015['nkill'].fillna(median_nkill, inplace=True)
data_2015['nwound'].fillna(median_nwound, inplace=True)

# Verificar si hay valores nulos restantes
print(data_2015[['nkill', 'nwound']].isnull().sum())

In [ ]:
data_2015.head()

In [ ]:
data_2015.shape

Ventajas de Usar la Mediana para Imputación en datos atípicos:

***Robustez frente a Valores Atípicos:

La mediana es menos sensible a los valores atípicos que la media. En otras palabras, los valores extremadamente altos o bajos no afectan la mediana tanto como afectan la media. Esto la hace una mejor opción para datasets que pueden contener valores atípicos, ya que la mediana proporcionará una estimación central más representativa de los datos subyacentes.

***Distribuciones Asimétricas:

En distribuciones asimétricas (sesgadas), la mediana proporciona una medida central más precisa que la media, que puede estar sesgada por los valores extremos. Por lo tanto, la mediana es más representativa del "centro" de los datos en estos casos.

***Simetría No Necesaria:

No se requiere que los datos sigan una distribución normal para que la mediana sea una buena medida central. Esto es particularmente útil en la práctica, donde muchas variables no siguen distribuciones normales.

---Imputación de Datos Perdidos:

Cuando se trata de imputar datos perdidos (missing data), hay varias opciones, y la elección de la mediana puede ser ventajosa en muchas situaciones:

***Simple y Eficaz:

La imputación con la mediana es un método simple y fácil de implementar. No requiere modelos complejos ni supuestos específicos sobre la distribución de los datos.

***Preservación de la Distribución Original:

Aunque la mediana no preserva la variabilidad de los datos imputados tan bien como métodos más complejos (como la imputación múltiple o los modelos de regresión), tiene la ventaja de mantener la forma básica de la distribución original de los datos.

***Menor Sesgo:

Imputar valores perdidos con la mediana introduce menos sesgo que la imputación con la media, especialmente en presencia de valores atípicos

# Analisis Multivariable

In [ ]:
fig, axis = plt.subplots(8, 2, figsize=(14, 32))
# Lista de pares de variables para gráficos de dispersión y mapas de calor
pairs = [
    ('success_num', 'nkill'),
    ('suicide_num', 'nwound'),
    ('country_txt_num', 'natlty1_txt_num'),
    ('attacktype1_txt_num', 'gname_num'),
    ('weaptype1_txt_num', 'attacktype1_txt_num'),
    ('ishostkid_num', 'INT_LOG_num'),
    ('INT_IDEO_num', 'region_txt_num'),
    ('provstate_num', 'weaptype1_txt_num')
]

# Generar gráficos de dispersión y mapas de calor para cada par
for i, (var1, var2) in enumerate(pairs):
    sns.regplot(ax=axis[i, 0], data=data_2015, x=var1, y=var2)
    sns.heatmap(data_2015[[var1, var2]].corr(), annot=True, fmt=".2f", ax=axis[i, 1], cbar=False, vmin=-1, vmax=1, cmap='coolwarm')

plt.tight_layout()
plt.show()

In [ ]:
fig, axis = plt.subplots(figsize = (20, 15))

sns.heatmap(data_2015[['nkill', 'nwound', 'country_txt_num', 'region_txt_num',
       'provstate_num', 'city_num', 'success_num', 'suicide_num',
       'attacktype1_txt_num', 'targtype1_txt_num', 'natlty1_txt_num',
       'gname_num', 'weaptype1_txt_num', 'property_num', 'ishostkid_num',
       'INT_LOG_num', 'INT_IDEO_num']].corr(), annot = True, fmt = ".2f")

plt.tight_layout()

plt.show()

Hay numerosas variables con alta correlacion lo que podria incurrir en un problema de muticolinealidad por lo que de las variables relacionadas con la ubicacion unicamente eligiremos natly_txt, la nacionalidad del perpetrador y prescindiremos del resto pues está estrechamente relacionada la nacionalidad de éste con el lugar del ataque suicida; INT_LOG e INT_IDEO tambien aparecen con alta correlacion eligiendo INT_IDEO, y tambien eliminaremos isthostkid por la alta correlacion con el tipo de ataque llevado a cabo.

In [ ]:
data.to_csv('data_EDA.csv', index=False)

In [ ]:
# Sumar las columnas 'nkill' y 'nwound' para obtener 'totales'
data_2015['totales'] = data_2015['nkill'] + data_2015['nwound']

# Verificar que se haya creado correctamente la columna
print(data_2015[['nkill', 'nwound', 'totales']].head())

In [ ]:
data_2015.head()

In [ ]:
variables = ['totales', 'date',  'country_txt_num', 'region_txt_num',
             'provstate_num', 'city_num', 'success_num', 'attacktype1_txt_num',
             'targtype1_txt_num', 'natlty1_txt_num', 'gname_num',
             'weaptype1_txt_num', 'property_num', 'ishostkid_num',
             'INT_LOG_num', 'INT_IDEO_num']

# Herramientas de Clasificacion

In [ ]:
data_2015.shape

In [ ]:
print(data_2015.columns)

In [ ]:
arbol_2015 = data_2015.copy()

In [ ]:
print(arbol_2015.columns)

In [ ]:
arbol_2015.drop(columns=['nkill', 'nwound'], inplace=True)


In [ ]:
print(arbol_2015.columns)

In [ ]:
# Lista de columnas a eliminar
columns_to_drop = ['country_txt_num',  'region_txt_num', 'provstate_num',
       'city_num', 'success_num', 'weaptype1_txt_num', 'ishostkid_num', 'INT_LOG_num']

arbol_2015 = arbol_2015.drop(columns=columns_to_drop)


In [ ]:
print(arbol_2015.columns)

In [ ]:
# Calcular la matriz de correlación
correlation_matrix = arbol_2015[[ 'suicide_num', 'attacktype1_txt_num',
       'targtype1_txt_num', 'natlty1_txt_num', 'gname_num', 'property_num',
       'INT_IDEO_num', 'totales']].corr()

# Visualizar la matriz de correlación
plt.figure(figsize=(12, 10))
sns.heatmap(correlation_matrix, annot=True, cmap='coolwarm', fmt=".2f")
plt.title('Matriz de Correlación')
plt.show()

In [ ]:

from sklearn.feature_selection import SelectKBest
from sklearn.model_selection import train_test_split

X = arbol_2015.drop("suicide_num", axis = 1)
y = arbol_2015["suicide_num"]

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 42)

selection_model = SelectKBest(k = 6)
selection_model.fit(X_train, y_train)

selected_columns = X_train.columns[selection_model.get_support()]
X_train_sel = pd.DataFrame(selection_model.transform(X_train), columns = selected_columns)
X_test_sel = pd.DataFrame(selection_model.transform(X_test), columns = selected_columns)

X_train_sel.head()

In [ ]:
X_test_sel.head()

# 3.BOOSTING

In [ ]:
from sklearn.ensemble import RandomForestClassifier, GradientBoostingClassifier
from xgboost import XGBClassifier
from sklearn.metrics import classification_report, confusion_matrix
from sklearn.tree import DecisionTreeClassifier, DecisionTreeRegressor
from sklearn.ensemble import RandomForestClassifier, RandomForestRegressor
from sklearn.ensemble import GradientBoostingClassifier, GradientBoostingRegressor
import xgboost as xgb
from xgboost import XGBClassifier, XGBRegressor




In [ ]:


# Supresión de advertencias
def warn(*args, **kwargs):
    pass

import warnings
warnings.warn = warn

In [ ]:
# Modelo de Árbol de Decisión
model = DecisionTreeClassifier(random_state=42)
model.fit(X_train, y_train)

# Predicciones
dt_pred = model.predict(X_test)

# Evaluación
print("Decision Tree Classifier:")
print(classification_report(y_test, dt_pred))
print("Confusion Matrix:\n", confusion_matrix(y_test, dt_pred))

In [ ]:
# Modelo de Bosque Aleatorio
rf_model = RandomForestClassifier(random_state=42)
rf_model.fit(X_train, y_train)

# Predicciones
rf_pred = rf_model.predict(X_test)

# Evaluación
print("Random Forest Classifier:")
print(classification_report(y_test, rf_pred))
print("Confusion Matrix:\n", confusion_matrix(y_test, rf_pred))

In [ ]:
# Modelo de Gradient Boosting
gb_model = GradientBoostingClassifier(random_state=42)
gb_model.fit(X_train, y_train)

# Predicciones
gb_pred = gb_model.predict(X_test)

# Evaluación
print("Gradient Boosting Classifier:")
print(classification_report(y_test, gb_pred))
print("Confusion Matrix:\n", confusion_matrix(y_test, gb_pred))

In [ ]:
from sklearn.model_selection import GridSearchCV
# Modelo de XGBoost
xgb_model = XGBClassifier(use_label_encoder=False, eval_metric='mlogloss', random_state=42)
param_grid = {
    'learning_rate': [0.01, 0.05, 0.1, 0.3],
    'n_estimators': [100, 200, 300],
    'max_depth': [3, 4, 5],
    'subsample': [0.6, 0.8, 1.0]
}

grid_search = GridSearchCV(estimator=xgb_model, param_grid=param_grid, scoring='accuracy', cv=5, n_jobs=-1)
grid_search.fit(X_train, y_train)

# Mejor estimador
best_xgb = grid_search.best_estimator_

# Predicciones
xgb_pred = best_xgb.predict(X_test)

# Evaluación
print("XGBoost Classifier:")
print(classification_report(y_test, xgb_pred))
print("Confusion Matrix:\n", confusion_matrix(y_test, xgb_pred))
print(f"Best hyperparameters: {grid_search.best_params_}")

XGBoost Classifier es el mejor modelo:

Precision:

Clase 0: 0.97 Clase 1: 0.69

Recall:

Clase 0: 0.98 Clase 1: 0.59

F1-Score:

Clase 0: 0.98 Clase 1: 0.63

Accuracy: 0.95

Matriz de Confusión:

Verdaderos negativos (Clase 0 predicha correctamente): 7224

Falsos positivos (Clase 1 predicha incorrectamente): 141

Falsos negativos (Clase 0 predicha incorrectamente): 217

Verdaderos positivos (Clase 1 predicha correctamente): 309

 Los altos valores de precisión y recall para la clase 0 indican que el modelo es efectivo al predecir cuando no hay suicidio involucrado. Esto es crucial, ya que ayuda a identificar correctamente los incidentes que no resultan en suicidio, lo cual puede ser útil para la planificación de la seguridad y la prevención.La menor precisión y recall para la clase 1 sugieren que el modelo podría beneficiarse de ajustes adicionales para mejorar la capacidad de detectar casos reales de suicidio. Esto podría incluir la exploración de técnicas de  recopilación de más datos relevantes (incluyendo por ejemplo mas años o alguno significativo como 2011) o ajuste de características del modelo (analizar con mas detenemiento y decidir si eliminar o añadir columnas).
 Weighted Avg: Es particularmente útil cuando hay desequilibrio de clases, como en este caso donde la clase 0 (no suicide) es mucho más frecuente que la clase 1 (suicide). Muestra cómo se desempeña el modelo en general, ponderando más las métricas en función del tamaño de cada clase, y en ese sentido se puede confirmar que es un buen analisis con un valor del 95%

In [ ]:
best_xgb = xgb.XGBRegressor(objective='reg:squarederror', n_estimators=100, random_state=42)
best_xgb.fit(X_train, y_train)


importancia = best_xgb.feature_importances_

# Crear un DataFrame para mostrar las importancias junto con los nombres de las variables
importancia_df = pd.DataFrame({'Feature': X_train.columns, 'Importance': importancia})

# Ordenar el DataFrame por importancia descendente
importancia_df = importancia_df.sort_values(by='Importance', ascending=False)

# Visualizar la importancia de las variables
plt.figure(figsize=(10, 8))
sns.barplot(x='Importance', y='Feature', data=importancia_df, palette='viridis')
plt.xlabel('Importancia')
plt.ylabel('Variables')
plt.title('Importancia de las Características')
plt.show()



Se podrian incluir todas, selectkbest no añade la ultima(daños a la propiedad), por lo que es correcto. Se puede apreciar como los ataques terroristas suicidas estan estrechamente relacionados con motivos ideologicos, el numero total de victimas, el tipo de ataque, el nombre del grupo terrorista que lo realiza, la nacionalidad de éste o éstos y el objetivo al que se dirige el ataque